From 40f06123191e47f493d05d65b19bd0efd37e3abc Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 3 Dec 2024 15:33:04 +0800
Subject: [PATCH 085/107] mtk: mac80211: refactor struct cfg80211_ttlm_params

1. add is_teardown in cfg80211_ttlm_params. If the "is_teardown" is
   set to true, the bitmap shoule become the default one and the
   dlink/ulink are meanless.
2. use IEEE80211_TTLM_NUM_TIDS for dlink/ulink array size.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 include/linux/ieee80211.h | 1 +
 include/net/cfg80211.h    | 7 +++++--
 include/net/mac80211.h    | 2 --
 3 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/include/linux/ieee80211.h b/include/linux/ieee80211.h
index 7b883c56..246c2358 100644
--- a/include/linux/ieee80211.h
+++ b/include/linux/ieee80211.h
@@ -251,6 +251,7 @@ static inline u16 ieee80211_sn_sub(u16 sn1, u16 sn2)
 
 #define IEEE80211_FIRST_TSPEC_TSID	8
 #define IEEE80211_NUM_TIDS		16
+#define IEEE80211_TTLM_NUM_TIDS		8
 
 /* number of user priorities 802.11 uses */
 #define IEEE80211_NUM_UPS		8
diff --git a/include/net/cfg80211.h b/include/net/cfg80211.h
index abb090b6..5bf7cdff 100644
--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -1741,14 +1741,17 @@ struct link_station_del_parameters {
  *
  * Used for setting a TID to link mapping.
  *
+ * @is_teardown: If it is a teardown request, the mapping should become the
+ * 	default one, and the dlink & ulink is meanless.
  * @dlink: Downlink TID to link mapping, as defined in section 9.4.2.314
  *     (TID-To-Link Mapping element) in Draft P802.11be_D4.0.
  * @ulink: Uplink TID to link mapping, as defined in section 9.4.2.314
  *     (TID-To-Link Mapping element) in Draft P802.11be_D4.0.
  */
 struct cfg80211_ttlm_params {
-	u16 dlink[8];
-	u16 ulink[8];
+	bool is_teardown;
+	u16 dlink[IEEE80211_TTLM_NUM_TIDS];
+	u16 ulink[IEEE80211_TTLM_NUM_TIDS];
 };
 
 /**
diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index 614331ed..e06fb621 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -1944,8 +1944,6 @@ struct ieee80211_vif_cfg {
 	u8 ap_addr[ETH_ALEN] __aligned(2);
 };
 
-#define IEEE80211_TTLM_NUM_TIDS 8
-
 /**
  * struct ieee8021_adv_ttlm - advertised TID to link map info
  *
-- 
2.45.2

